// apollo-server-express/dist/ApolloServer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cors_1 = __importDefault(require("cors"));
const body_parser_1 = require("body-parser");
const graphql_playground_middleware_express_1 = __importDefault(require("graphql-playground-middleware-express"));
const apollo_server_core_1 = require("apollo-server-core");
const accepts_1 = __importDefault(require("accepts"));
const type_is_1 = __importDefault(require("type-is"));
const expressApollo_1 = require("./expressApollo");
const apollo_upload_server_1 = require("apollo-upload-server");
var apollo_server_core_2 = require("apollo-server-core");
exports.GraphQLExtension = apollo_server_core_2.GraphQLExtension;
const apollo_server_core_3 = require("apollo-server-core");
const fileUploadMiddleware = (uploadsConfig, server) => (req, res, next) => {
    if (type_is_1.default(req, ['multipart/form-data'])) {
        apollo_upload_server_1.processRequest(req, uploadsConfig)
            .then(body => {
            req.body = body;
            next();
        })
            .catch(error => {
            if (error.status && error.expose)
                res.status(error.status);
            next(apollo_server_core_1.formatApolloErrors([error], {
                formatter: server.requestOptions.formatError,
                debug: server.requestOptions.debug,
                logFunction: server.requestOptions.logFunction,
            }));
        });
    }
    else {
        next();
    }
};
class ApolloServer extends apollo_server_core_1.ApolloServerBase {
    createGraphQLServerOptions(req, res) {
        const _super = name => super[name];
        return __awaiter(this, void 0, void 0, function* () {
            return _super("graphQLServerOptions").call(this, { req, res });
        });
    }
    supportsSubscriptions() {
        return true;
    }
    applyMiddleware({ app, path, cors, bodyParserConfig, disableHealthCheck, gui, onHealthCheck, uploads, }) {
        if (!path)
            path = '/graphql';
        if (!disableHealthCheck) {
            app.use('/.well-known/apollo/server-health', (req, res) => {
                res.type('application/health+json');
                if (onHealthCheck) {
                    onHealthCheck(req)
                        .then(() => {
                        res.json({ status: 'pass' });
                    })
                        .catch(() => {
                        res.status(503).json({ status: 'fail' });
                    });
                }
                else {
                    res.json({ status: 'pass' });
                }
            });
        }
        let uploadsMiddleware;
        if (uploads !== false) {
            this.enhanceSchema(apollo_server_core_3.makeExecutableSchema({
                typeDefs: apollo_server_core_3.gql `
            scalar Upload
          `,
                resolvers: { Upload: apollo_upload_server_1.GraphQLUpload },
            }));
            uploadsMiddleware = fileUploadMiddleware(typeof uploads !== 'boolean' ? uploads : {}, this);
        }
        this.graphqlPath = path;
        if (cors === true) {
            app.use(path, cors_1.default());
        }
        else if (cors !== false) {
            app.use(path, cors_1.default(cors));
        }
        if (bodyParserConfig === true) {
            app.use(path, body_parser_1.json());
        }
        else if (bodyParserConfig !== false) {
            app.use(path, body_parser_1.json(bodyParserConfig));
        }
        if (uploadsMiddleware) {
            app.use(path, uploadsMiddleware);
        }
        const guiEnabled = !!gui || (gui === undefined && process.env.NODE_ENV !== 'production');
        app.use(path, (req, res, next) => {
            if (guiEnabled && req.method === 'GET') {
                const accept = accepts_1.default(req);
                const types = accept.types();
                const prefersHTML = types.find((x) => x === 'text/html' || x === 'application/json') === 'text/html';
                if (prefersHTML) {
                    const middlewareOptions = Object.assign({ endpoint: path, subscriptionEndpoint: this.subscriptionsPath }, (typeof gui === 'boolean' ? {} : gui));
                    return graphql_playground_middleware_express_1.default(middlewareOptions)(req, res, next);
                }
            }
            return expressApollo_1.graphqlExpress(this.createGraphQLServerOptions.bind(this))(req, res, next);
        });
    }
}
exports.ApolloServer = ApolloServer;
exports.registerServer = () => {
    throw new Error('Please use server.applyMiddleware instead of registerServer. This warning will be removed in the next release');
};
